### PR TITLE
API cleanup standardise error codes and types

### PR DESCRIPTION
**Pull request by despo**
## Changes proposed in this pull request

- Changed all unprocessable entity errors to return `UnprocessableEntityResponse` as part of error JSON for consistency
- Introduced new matches that  checks for errors name and message for simplicity 
- Changed unauthorised error to return `Unauthorised` instead of `NotAuthorised` to match API docs
- Removed custom validation errors
- Reuse APIValidationsAndErrorHandling across all VendorAPI controllers
- Improved example wording in relation to expected errors and tested against specific API versions

## Link to Trello card

https://trello.com/c/Ag22HqdJ/4816-api-cleanup-standardise-error-codes-and-types

## Things to check

- [x] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] Required environment variables have been updated [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

### CODE REVIEW

File: app/controllers/vendor_api/decisions_controller.rb [hunk 1]
```diff
@@ -89,7 +89,7 @@ def respond_to_decision(decision)
 
     # Takes errors from ActiveModel::Validations and render them in the API response
     def render_validation_errors(errors)
-      error_responses = errors.full_messages.map { |message| { error: 'ValidationError', message: message } }
+      error_responses = errors.full_messages.map { |message| { error: 'UnprocessableEntity', message: message } }
```
Comment (cwrw) [inline app/controllers/vendor_api/decisions_controller.rb]: todo: do not change v1 errors

File: spec/requests/vendor_api/v1.0/post_conditions_met_spec.rb [hunk 2]
```diff
@@ -36,7 +36,7 @@
     expect(response).to have_http_status(:not_found)
     expect(parsed_response)
       .to contain_schema_with_error('NotFoundResponse',
-                                    'Could not find an application with ID non-existent-id',
+                                    'Unable to find Application(s)',
```
Comment (cwrw) [inline spec/requests/vendor_api/v1.0/post_conditions_met_spec.rb:39]: check with vendors around this

**Review by cwrw — COMMENTED** [review]
great cleanup! need to check around v1 validations with vendors after that we can commit to this, also need to update v1.1/draft spec

File: app/controllers/vendor_api/decisions_controller.rb (hunk 1 above)
Comment (cwrw) [reply app/controllers/vendor_api/decisions_controller.rb]: I double checked today and should be ok to change those

**Review by cwrw — COMMENTED** [review]
✅ looks good, @SamJamCul let's have a look tomorrow

File: app/controllers/vendor_api/interviews_controller.rb [hunk 3]
```diff
@@ -46,27 +44,21 @@ def cancel
...
-        errors: [
-          {
-            error: e.class.to_s,
-            message: e.message,
-          },
-        ],
-      }
+    def handle_as_validation_error(e)
+      render status: :unprocessable_entity,
+             json: { errors: [{ error: 'UnprocessableEntityResponse', message: e.message }] }
```
Comment (SamJamCul) [inline app/controllers/vendor_api/interviews_controller.rb:48]: Should this just be `UnprocessableEntity` only

File: spec/requests/vendor_api/v1.0/get_multiple_applications_spec.rb [hunk 4]
```diff
@@ -52,44 +52,41 @@
...
     get_api_request '/api/v1/applications'
 
     expect(response).to have_http_status(:unprocessable_entity)
-
-    expect(parsed_response).to be_valid_against_openapi_schema('ParameterMissingResponse')
-
-    expect(error_response['message']).to eql('param is missing or the value is empty: since')
+    expect(parsed_response).to contain_schema_with_error('ParameterMissingResponse',
+                                                         'param is missing or the value is empty: since',
+                                                         '1.0')
```
Comment (SamJamCul) [inline spec/requests/vendor_api/v1.0/get_multiple_applications_spec.rb:61]: Remove reference to 1.0

**Review by cwrw — APPROVED** [review]

**Pull request merged**